### PR TITLE
Feature csi 3447 nvme controller

### PR DESCRIPTION
**Pull request by ArbelNathan**

### CODE REVIEW

File: controller/array_action/array_mediator_interface.py [hunk 1]
```diff
@@ -151,13 +151,14 @@ def get_volume_mappings(self, volume_id):
...
     @abstractmethod
-    def map_volume(self, volume_id, host_name):
+    def map_volume(self, volume_id, host_name, connectivity_type):
         """
         This function will find the next available lun for the host and map the volume to it.
 
         Args:
            volume_id : the volume WWN.
            host_name : the name of the host to map the volume to.
+           connectivity_type : connectivity type (iscsi, fc or nvme)
```
Comment (oriyarde) [inline controller/array_action/array_mediator_interface.py:161]: I would avoid listing examples in a way which sounds exhaustive. consider something like:
```suggestion
           connectivity_type : connectivity type (e.g. iscsi, fc, etc...)
```

Comment (ArbelNathan) [reply controller/array_action/array_mediator_interface.py:161]: those are not examples, they are the only options we support.

File: controller/array_action/array_mediator_svc.py [hunk 2]
```diff
@@ -626,29 +634,32 @@ def delete_snapshot(self, snapshot_id):
...
             raise array_errors.MultipleHostsFoundError(initiators, fc_host)
-        if iscsi_host:
-            logger.debug("found host : {0} with iqn : {1}".format(iscsi_host, initiators.iscsi_iqn))
-            return iscsi_host, [config.ISCSI_CONNECTIVITY_TYPE]
-        if fc_host:
-            logger.debug("found host : {0} with fc wwn : {1}".format(fc_host, initiators.fc_wwns))
-            return fc_host, [config.FC_CONNECTIVITY_TYPE]
-        logger.debug("can not found host by using initiators: {0} ".format(initiators))
-        raise array_errors.HostNotFoundError(initiators)
+        return host_name, list(connectivity_types)
```
Comment (oriyarde) [inline controller/array_action/array_mediator_svc.py:662]: why use `set` if we return a `list`?

File: controller/array_action/array_mediator_interface.py (hunk 1 above)
Comment (oriyarde) [reply controller/array_action/array_mediator_interface.py:161]: I don't think we should maintain all the options we support here

File: controller/array_action/array_mediator_svc.py [hunk 3]
```diff
@@ -761,8 +773,11 @@ def map_volume(self, volume_id, host_name):
         }
 
         try:
-            lun = self.get_first_free_lun(host_name)
-            cli_kwargs.update({'scsi': lun})
+            if connectivity_type != config.NVME_OVER_FC_CONNECTIVITY_TYPE:
+                lun = self.get_first_free_lun(host_name)
+                cli_kwargs.update({'scsi': lun})
+            else:
+                lun = config.NVME_OVER_FC_CONNECTIVITY_TYPE
```
Comment (oriyarde) [inline controller/array_action/array_mediator_svc.py:780]: why is this needed?

File: controller/array_action/array_mediator_svc.py (hunk 2 above)
Comment (ArbelNathan) [reply controller/array_action/array_mediator_svc.py:662]: duplicates (multiple hosts with the same connectivity type)

File: controller/array_action/array_mediator_svc.py (hunk 3 above)
Comment (ArbelNathan) [reply controller/array_action/array_mediator_svc.py:780]: cause nvme is not using lun

Comment (oriyarde) [reply controller/array_action/array_mediator_svc.py:780]: I still don't understand why line 780 here is needed

Comment (ArbelNathan) [reply controller/array_action/array_mediator_svc.py:780]: so there will be an indication for it in the logs instead of just None...

Comment (oriyarde) [reply controller/array_action/array_mediator_svc.py:780]: besides, I see the lun being logged later with the connectivity type, so to me the logs should be fine

File: controller/array_action/array_mediator_svc.py [hunk 4]
```diff
@@ -1087,3 +1102,10 @@ def demote_replication_volume(self, replication_name):
         rcrelationship = self._get_rcrelationship_by_name(replication_name)
         endpoint_type_to_promote = self._get_replication_other_endpoint_type(rcrelationship)
         self._ensure_endpoint_is_primary(rcrelationship, endpoint_type_to_promote)
+
+    def get_host_name_if_equal(self, iscsi_host, nvme_host, fc_host):
+        s = {iscsi_host, nvme_host, fc_host}
+        s.discard(None)
+        if len(s) > 1:
+            return None
+        return s.pop()
```
Comment (oriyarde) [inline controller/array_action/array_mediator_svc.py:1111]: in this implementation, `get_host_name_if_equal` doesn't support all-None parameters (even though it won't happen with current `get_host_by_host_identifiers` implementation).
```suggestion
        if len(s) == 1:
            return s.pop()
        return None
```

File: controller/array_action/utils.py [hunk 5]
```diff
@@ -1,11 +1,30 @@
...
+def convert_scsi_uuid_to_nguid(vol_id):
+    logger.debug("converting scsi uuid to nguid")
+    indices = [1, 7, 16]
+    splited_text = list()
+    for index, _ in enumerate(indices):
+        last = len(vol_id)
+        if index + 1 < len(indices):
+            last = indices[index + 1]
+        splited_text.append(vol_id[indices[index]:last])
+    final_nguid = splited_text[2] + splited_text[0] + '0' + splited_text[1]
```
Comment (oriyarde) [inline controller/array_action/utils.py:23]: could we have names for each parsed part, instead of just hardcoded numbers?
also `indices[index]` is confusing.
could you write an example here for a string, the parts to which it needs to be broken down, and how they are composed to the final value?
I'm sure we can find good names for every variable

File: controller/array_action/array_mediator_svc.py (hunk 2 above)
Comment (ArbelNathan) [reply controller/array_action/array_mediator_svc.py:662]: actually, I dont sure it's possible, I think its a UT error

File: controller/common/node_info.py [hunk 6]
```diff
@@ -52,5 +55,16 @@ def is_array_iscsi_iqns_match(self, host_iqns):
...
+    def is_array_nvme_nqn_match(self, host_nqns):
+        """
+        Args:
+           host_nqns: storage host nqns list
+
+        Returns:
+           Is current host nqns matches
+        """
+        host_nqns_lower = [nqn.lower() for nqn in host_nqns]
+        return self._nvme_nqn_lowercase in host_nqns_lower
```
Comment (oriyarde) [inline controller/common/node_info.py:67]: could we just rename `is_array_iscsi_iqns_match`/move this to a common function?

File: controller/common/node_info.py [hunk 7]
```diff
@@ -52,5 +55,16 @@ def is_array_iscsi_iqns_match(self, host_iqns):
...
+
     def __str__(self):
-        return "iscsi_iqn: " + self.iscsi_iqn + ", fc_wwns: " + ",".join(self.fc_wwns)
+        return "iscsi_iqn: " + self.iscsi_iqn + ", fc_wwns: " + ",".join(self.fc_wwns) + "nvme_nqn: " + self.nvme_nqn
```
Comment (oriyarde) [inline controller/common/node_info.py:70]: consider choosing a consistent order for the connectivity types wherever applicable

File: controller/tests/array_action/svc/array_mediator_svc_test.py [hunk 8]
```diff
@@ -621,70 +629,62 @@ def _prepare_mocks_for_get_host_by_identifiers_(self, result_reader_iter, hosts)
...
     def test_get_host_by_identifiers_returns_host_not_found(self, result_reader_iter):
-        host_1 = self._get_host_as_dictionary('host_id_1', 'test_host_1', ['iqn.test.1'], [])
-        host_2 = self._get_host_as_dictionary('host_id_2', 'test_host_2', ['iqn.test.2'], [])
-        host_3 = self._get_host_as_dictionary('host_id_3', 'test_host_3', ['iqn.test.3'], [])
-        hosts = [host_1, host_2, host_3]
-        self._prepare_mocks_for_get_host_by_identifiers_(result_reader_iter, hosts)
+        self._prepare_mocks_for_get_host_by_identifiers_(result_reader_iter)
         with self.assertRaises(array_errors.HostNotFoundError):
-            self.svc.get_host_by_host_identifiers(Initiators('Test_iqn', ['Test_wwn']))
+            self.svc.get_host_by_host_identifiers(Initiators('Test_iqn', ['Test_WWN'], ''))
```
Comment (oriyarde) [inline controller/tests/array_action/svc/array_mediator_svc_test.py:634]: I think it's usually:
```suggestion
            self.svc.get_host_by_host_identifiers(Initiators('Test_iqn', ['Test_wwn'], ''))
```
any reason not to keep it aligned?

File: controller/tests/array_action/svc/array_mediator_svc_test.py [hunk 9]
```diff
@@ -694,33 +694,33 @@ def test_get_host_by_identifiers_with_wrong_fc_iscsi_raise_not_found(self, resul
...
-        host_1 = self._get_host_as_dictionary('host_id_1', 'test_host_1', [], ['abc1'])
-        host_2 = self._get_host_as_dictionary('host_id_2', 'test_host_2', ['iqn.test.6'], ['abcd3'])
-        host_3 = self._get_host_as_dictionary('host_id_3', 'test_host_3', ['iqn.test.2'], ['abc3'])
-        hosts = [host_1, host_2, host_3]
-        self._prepare_mocks_for_get_host_by_identifiers_(result_reader_iter, hosts)
-        host, connectivity_type = self.svc.get_host_by_host_identifiers(Initiators('iqn.test.2', ['ABC3']))
-        self.assertEqual('test_host_3', host)
-        self.assertEqual([config.ISCSI_CONNECTIVITY_TYPE,
-                          config.FC_CONNECTIVITY_TYPE], connectivity_type)
+    def test_get_host_by_identifiers_return_iscsi_nvme_fc_all_support(self, result_reader_iter):
```
Comment (oriyarde) [inline controller/tests/array_action/svc/array_mediator_svc_test.py:705]: I would remove `_all_support`

File: controller/array_action/array_mediator_svc.py (hunk 2 above)
Comment (oriyarde) [reply controller/array_action/array_mediator_svc.py:662]: is it not possible to have multiple hosts with the same connectivity type (user misconfiguration)?

File: controller/common/node_info.py (hunk 6 above)
Comment (oriyarde) [reply controller/common/node_info.py:67]: true. but can't we pass it as a parameter?

Comment (ArbelNathan) [reply controller/common/node_info.py:67]: sure

Comment (ArbelNathan) [reply controller/common/node_info.py:67]: @lisyukanna

File: controller/tests/array_action/svc/array_mediator_svc_test.py (hunk 8 above)
Comment (ArbelNathan) [reply controller/tests/array_action/svc/array_mediator_svc_test.py]: done

File: controller/array_action/array_mediator_svc.py [hunk 10]
```diff
@@ -1087,3 +1102,10 @@ def demote_replication_volume(self, replication_name):
         rcrelationship = self._get_rcrelationship_by_name(replication_name)
         endpoint_type_to_promote = self._get_replication_other_endpoint_type(rcrelationship)
         self._ensure_endpoint_is_primary(rcrelationship, endpoint_type_to_promote)
+
+    def get_host_name_if_equal(self, iscsi_host, nvme_host, fc_host):
```
Comment (oriyarde) [inline controller/array_action/array_mediator_svc.py:1106]: why is this method public?

File: controller/common/node_info.py (hunk 7 above)
Comment (oriyarde) [reply controller/common/node_info.py]: I mean [throughout](https://github.com/IBM/ibm-block-csi-driver/blob/574a573978764d487dd86c8e1581b04f32e86cc6/controller/common/node_info.py#L22) [the](https://github.com/IBM/ibm-block-csi-driver/blob/574a573978764d487dd86c8e1581b04f32e86cc6/controller/array_action/array_mediator_svc.py#L637) [PR](https://github.com/IBM/ibm-block-csi-driver/blob/574a573978764d487dd86c8e1581b04f32e86cc6/controller/array_action/array_mediator_svc.py#L1106)...

File: controller/array_action/array_mediator_svc.py (hunk 10 above)
Comment (ArbelNathan) [reply controller/array_action/array_mediator_svc.py]: right.
done

File: controller/tests/array_action/svc/array_mediator_svc_test.py [hunk 11]
```diff
@@ -621,70 +629,62 @@ def _prepare_mocks_for_get_host_by_identifiers_(self, result_reader_iter, hosts)
...
     def test_get_host_by_identifiers_raise_multiplehostsfounderror(self, result_reader_iter):
-        host_1 = self._get_host_as_dictionary('host_id_1', 'test_host_1', ['iqn.test.1'], [])
-        host_2 = self._get_host_as_dictionary('host_id_2', 'test_host_2', ['iqn.test.3'], [])
-        host_3 = self._get_host_as_dictionary('host_id_3', 'test_host_3', [], ['Test_wwn'])
-        hosts = [host_1, host_2, host_3]
-        self._prepare_mocks_for_get_host_by_identifiers_(result_reader_iter, hosts)
+        self._prepare_mocks_for_get_host_by_identifiers_(result_reader_iter)
         with self.assertRaises(array_errors.MultipleHostsFoundError):
-            self.svc.get_host_by_host_identifiers(Initiators('iqn.test.3', ['Test_wwn']))
+            self.svc.get_host_by_host_identifiers(Initiators('iqn.test.3', ['abc2'], ''))
```
[the remark this answers is not in the document]
Comment (oriyarde) [reply controller/tests/array_action/svc/array_mediator_svc_test.py]: I still see this:
image

File: controller/common/node_info.py (hunk 7 above)
Comment (oriyarde) [reply controller/common/node_info.py]: then let's choose the order with minimum effort.
I don't think it matters what order, as long as it is consistent

File: controller/array_action/utils.py (hunk 5 above)
Comment (oriyarde) [reply controller/array_action/utils.py:23]: nit: I personally prefer `final_nguid = ''.join(vendor_identifier_extension, oui, '0', vendor_identifier`)

Comment (oriyarde) [reply controller/array_action/utils.py:23]: we can also consider using consts, e.g. `WWN_OUI_END = 7`

File: controller/common/node_info.py (hunk 7 above)
Comment (ArbelNathan) [reply controller/common/node_info.py]: done

File: controller/array_action/array_mediator_svc.py [hunk 12]
```diff
@@ -12,7 +12,7 @@
 from controller.array_action.array_action_types import Volume, Snapshot, Host, Replication
 from controller.array_action.array_mediator_abstract import ArrayMediatorAbstract
 from controller.array_action.svc_cli_result_reader import SVCListResultsReader
-from controller.array_action.utils import classproperty, bytes_to_string
+from controller.array_action.utils import classproperty, bytes_to_string, convert_scsi_uuid_to_nguid
```
[the remark this answers is not in the document]
Comment (ArbelNathan) [reply controller/array_action/array_mediator_svc.py]: done

File: controller/array_action/utils.py (hunk 5 above)
Comment (ArbelNathan) [reply controller/array_action/utils.py]: done

File: controller/tests/array_action/svc/array_mediator_svc_test.py [hunk 13]
```diff
@@ -813,7 +852,7 @@ def test_first_free_lun_no_available_lun(self):
     @patch("controller.array_action.array_mediator_svc.SVCArrayMediator.get_first_free_lun")
     def _test_map_volume_mkvdiskhostmap_error(self, client_error, expected_error, mock_get_first_free_lun):
         mock_get_first_free_lun.return_value = '1'
-        self._test_mediator_method_client_error(self.svc.map_volume, ("volume", "host"),
+        self._test_mediator_method_client_error(self.svc.map_volume, ("volume", "host", ""),
```
[the remark this answers is not in the document]
Comment (ArbelNathan) [reply controller/tests/array_action/svc/array_mediator_svc_test.py]: done

File: controller/tests/controller_server/utils_test.py [hunk 14]
```diff
@@ -383,51 +385,77 @@ def test_get_volume_id_info_no_id_fail(self):
...
+
+        res = utils.choose_connectivity_type([ISCSI_CONNECTIVITY_TYPE])
+        self.assertEqual(res, ISCSI_CONNECTIVITY_TYPE)
+
+        res = utils.choose_connectivity_type(
+            [NVME_OVER_FC_CONNECTIVITY_TYPE, FC_CONNECTIVITY_TYPE, ISCSI_CONNECTIVITY_TYPE])
+        self.assertEqual(res, NVME_OVER_FC_CONNECTIVITY_TYPE)
+
+        res = utils.choose_connectivity_type([FC_CONNECTIVITY_TYPE, ISCSI_CONNECTIVITY_TYPE])
+        self.assertEqual(res, FC_CONNECTIVITY_TYPE)
```
[the remark this answers is not in the document]
Comment (ArbelNathan) [reply controller/tests/controller_server/utils_test.py]: done

File: controller/tests/array_action/ds8k/test_array_mediator_ds8k.py [hunk 15]
```diff
@@ -303,28 +303,28 @@ def test_get_volume_mappings(self):
...
         with self.assertRaises(array_errors.MappingError):
-            self.array.map_volume("fake_name", "fake_host")
+            self.array.map_volume("fake_name", "fake_host", "fake_connectivity_type")
 
     def test_map_volume(self):
         scsi_id = "6005076306FFD3010000000000000001"
         host_name = "test_name"
         self.client_mock.map_volume_to_host.return_value = Munch({"lunid": "01"})
-        lun = self.array.map_volume(scsi_id, host_name)
+        lun = self.array.map_volume(scsi_id, host_name, "fake_connectivity_type")
```
Comment (oriyarde) [inline controller/tests/array_action/ds8k/test_array_mediator_ds8k.py:327]: it seems in the function the argument values should all be defined in variables before the call

File: controller/tests/array_action/svc/array_mediator_svc_test.py [hunk 16]
```diff
@@ -621,106 +632,130 @@ def _prepare_mocks_for_get_host_by_identifiers_(self, result_reader_iter, hosts)
...
-            'iqn.test.6', ['abc3']))
-        self.assertEqual('test_host_3', host)
-        self.assertEqual([config.FC_CONNECTIVITY_TYPE], connectivity_type)
+        hostname, connectivity_types = self.svc.get_host_by_host_identifiers(
+            Initiators('Test_nqn', ['wwn3'], 'iqn.test.6'))
+        self.assertEqual('test_host_3', hostname)
+        self.assertEqual([config.FC_CONNECTIVITY_TYPE], connectivity_types)
+
+    @patch("controller.array_action.svc_cli_result_reader.SVCListResultsReader.__iter__")
+    def test_get_host_by_identifiers_return_fc_host_no_other_ports(self, result_reader_iter):
```
[the remark this answers is not in the document]
Comment (ArbelNathan) [reply controller/tests/array_action/svc/array_mediator_svc_test.py]: done

File: controller/tests/array_action/ds8k/test_array_mediator_ds8k.py (hunk 15 above)
Comment (ArbelNathan) [reply controller/tests/array_action/ds8k/test_array_mediator_ds8k.py]: done

File: controller/tests/array_action/svc/array_mediator_svc_test.py [hunk 17]
```diff
@@ -621,106 +632,130 @@ def _prepare_mocks_for_get_host_by_identifiers_(self, result_reader_iter, hosts)
...
+    def test_get_host_by_identifiers_return_nvme_host(self, result_reader_iter):
+        self._prepare_mocks_for_get_host_by_identifiers(result_reader_iter)
+        hostname, connectivity_types = self.svc.get_host_by_host_identifiers(
+            Initiators('nqn.test.3', ['Test_wwn'], 'iqn.test.6'))
+        self.assertEqual('test_host_3', hostname)
+        self.assertEqual([config.NVME_OVER_FC_CONNECTIVITY_TYPE], connectivity_types)
+
+    @patch("controller.array_action.svc_cli_result_reader.SVCListResultsReader.__iter__")
+    def test_get_host_by_identifiers_no_other_ports_return_nvme_host(self, result_reader_iter):
+        host_with_iqn_list = self._get_host_as_dictionary('host_id_s', 'test_host_s', nqn_list=['nqn.test.s'])
```
Comment (oriyarde) [inline controller/tests/array_action/svc/array_mediator_svc_test.py:689]:
```suggestion
        host_with_nqn_list = self._get_host_as_dictionary('host_id_s', 'test_host_s', nqn_list=['nqn.test.s'])
```

File: controller/tests/array_action/svc/array_mediator_svc_test.py [hunk 18]
```diff
@@ -621,106 +632,130 @@ def _prepare_mocks_for_get_host_by_identifiers_(self, result_reader_iter, hosts)
...
     def test_get_host_by_identifiers_return_fc_host(self, result_reader_iter):
-        host_1 = self._get_host_as_dictionary('host_id_1', 'test_host_1', [], ['abc1'])
-        host_2 = self._get_host_as_dictionary('host_id_2', 'test_host_2', [''], ['abc2'])
-        host_3 = self._get_host_as_dictionary('host_id_3', 'test_host_3', ['iqn.test.3'], ['abc1', 'abc3'])
+        host_1 = self._get_host_as_dictionary('host_id_1', 'test_host_1', wwpns_list=['wwn1'], iscsi_names_list=[])
+        host_2 = self._get_host_as_dictionary('host_id_2', 'test_host_2', wwpns_list=['wwn2'], iscsi_names_list=[''])
+        host_3 = self._get_host_as_dictionary('host_id_3', 'test_host_3', wwpns_list=['wwn3', 'wwn4'],
+                                              iscsi_names_list=['iqn.test.3'])
         hosts = [host_1, host_2, host_3]
-        self._prepare_mocks_for_get_host_by_identifiers_(result_reader_iter, hosts)
```
Comment (oriyarde) [inline controller/tests/array_action/svc/array_mediator_svc_test.py:679]: so where is `hosts` being used now?

File: controller/common/node_info.py [hunk 19]
```diff
@@ -52,5 +55,16 @@ def is_array_iscsi_iqns_match(self, host_iqns):
...
+
+        Returns:
+           Is current host nqns matches
+        """
+        host_nqns_lower = [nqn.lower() for nqn in host_nqns]
+        return self._nvme_nqn_lowercase in host_nqns_lower
+
     def __str__(self):
-        return "iscsi_iqn: " + self.iscsi_iqn + ", fc_wwns: " + ",".join(self.fc_wwns)
+        return "nvme_nqn: " + self.nvme_nqn + ", fc_wwns: " + ",".join(self.fc_wwns) + " iscsi_iqn: " + self.iscsi_iqn
```
Comment (oriyarde) [inline controller/common/node_info.py:70]: not a must, but since it's starting to get long, I think format would be more readable

File: controller/tests/array_action/svc/array_mediator_svc_test.py [hunk 20]
```diff
@@ -621,106 +632,130 @@ def _prepare_mocks_for_get_host_by_identifiers_(self, result_reader_iter, hosts)
...
+        self.assertEqual('test_host_3', hostname)
+        self.assertEqual([config.FC_CONNECTIVITY_TYPE], connectivity_types)
+
+    @patch("controller.array_action.svc_cli_result_reader.SVCListResultsReader.__iter__")
+    def test_get_host_by_identifiers_no_other_ports_return_fc_host(self, result_reader_iter):
+        host_with_iqn_list = self._get_host_as_dictionary('host_id_s', 'test_host_s', wwpns_list=['WWNs'])
+        self._prepare_mocks_for_get_host_by_identifiers(result_reader_iter, special_host=host_with_iqn_list)
+        hostname, connectivity_types = self.svc.get_host_by_host_identifiers(
+            Initiators('Test_nqn', ['Test_wwn', 'WWNs'], 'Test_iqn'))
+        self.assertEqual('test_host_s', hostname)
```
Comment (oriyarde) [inline controller/tests/array_action/svc/array_mediator_svc_test.py:721]: is there a meaning to the `s` in `test_host_s` here?

File: controller/tests/array_action/svc/array_mediator_svc_test.py [hunk 21]
```diff
@@ -732,14 +767,18 @@ def _get_detailed_hosts_list_result(self, hosts_dict):
...
             current_element.add(HOST_ID_PARAM, host_dict.get(HOST_ID_PARAM))
             current_element.add(HOST_NAME_PARAM, host_dict.get(HOST_NAME_PARAM))
-            iscsi_names_list = host_dict.get(HOST_ISCSI_NAMES_PARAM)
-            if iscsi_names_list:
-                for iscsi_name in iscsi_names_list:
-                    current_element.add(HOST_ISCSI_NAMES_PARAM, iscsi_name)
+            nvme_nqn_list = host_dict.get(HOST_NQN_PARAM)
+            if nvme_nqn_list:
+                for nqn in nvme_nqn_list:
+                    current_element.add(HOST_NQN_PARAM, nqn)
```
Comment (oriyarde) [inline controller/tests/array_action/svc/array_mediator_svc_test.py:773]: not a must (it's a gap in existing code), but I think this can be better:
```suggestion
            nvme_nqn_list = host_dict.get(HOST_NQN_PARAM, [])
            for nqn in nvme_nqn_list:
                current_element.add(HOST_NQN_PARAM, nqn)
```

File: controller/tests/array_action/svc/array_mediator_svc_test.py (hunk 18 above)
Comment (ArbelNathan) [reply controller/tests/array_action/svc/array_mediator_svc_test.py:679]: it sent after in that method

File: controller/tests/array_action/svc/array_mediator_svc_test.py (hunk 21 above)
Comment (ArbelNathan) [reply controller/tests/array_action/svc/array_mediator_svc_test.py]: nice! done

File: controller/common/node_info.py (hunk 19 above)
Comment (ArbelNathan) [reply controller/common/node_info.py]: done

File: controller/tests/array_action/svc/array_mediator_svc_test.py (hunk 17 above)
Comment (ArbelNathan) [reply controller/tests/array_action/svc/array_mediator_svc_test.py]: yup, done

File: controller/tests/array_action/svc/array_mediator_svc_test.py (hunk 20 above)
Comment (oriyarde) [reply controller/tests/array_action/svc/array_mediator_svc_test.py:721]: so could we have an indicative name? e.g. `test_special_host`

File: controller/tests/array_action/svc/array_mediator_svc_test.py (hunk 17 above)
Comment (oriyarde) [reply controller/tests/array_action/svc/array_mediator_svc_test.py]: I still see this gap in other places (e.g. the sloppy link).
please don't rely only on my examples - I might miss some.
the best strategy is for you to first go over all places, and only then I will go over all places as well.

File: controller/tests/array_action/svc/array_mediator_svc_test.py (hunk 20 above)
Comment (ArbelNathan) [reply controller/tests/array_action/svc/array_mediator_svc_test.py:721]: I tried to keep aligned to the other names - test_host_1, test_host_2, test_host_s ...

Comment (oriyarde) [reply controller/tests/array_action/svc/array_mediator_svc_test.py:721]: though I would prefer a non-festive term, e.g. "custom"

File: controller/tests/controller_server/utils_test.py [hunk 22]
```diff
@@ -383,51 +385,70 @@ def test_get_volume_id_info_no_id_fail(self):
...
-        self.assertEqual(res, "fc")
+        node_id_info = NodeIdInfo("{};;{}".format(host_name, fc_wwns))
+        self._check_node_id_parameters(node_id_info, "", fc_wwns, "")
 
-        res = utils.choose_connectivity_type(["iscsi", "fc"])
-        self.assertEqual(res, "fc")
+    def test_choose_connectivity_types(self):
+        iscsi = ISCSI_CONNECTIVITY_TYPE
+        fc = FC_CONNECTIVITY_TYPE
+        nvme = NVME_OVER_FC_CONNECTIVITY_TYPE
```
Comment (oriyarde) [inline controller/tests/controller_server/utils_test.py:421]:
```suggestion
        nvme = NVME_OVER_FC_CONNECTIVITY_TYPE
        fc = FC_CONNECTIVITY_TYPE
        iscsi = ISCSI_CONNECTIVITY_TYPE
```
etc...

File: controller/controller_server/config.py [hunk 23]
```diff
@@ -47,7 +47,7 @@
 
 REQUEST_ACCESSIBILITY_REQUIREMENTS_FIELD = "accessibility_requirements"
 
-SUPPORTED_CONNECTIVITY_TYPES = 2
+SUPPORTED_CONNECTIVITY_TYPES = 3
```
[the remark this answers is not in the document]
Comment (ArbelNathan) [reply controller/controller_server/config.py]: we are not, I removed it

File: controller/tests/controller_server/utils_test.py (hunk 22 above)
Comment (ArbelNathan) [reply controller/tests/controller_server/utils_test.py]: done

File: controller/tests/array_action/svc/array_mediator_svc_test.py (hunk 20 above)
Comment (ArbelNathan) [reply controller/tests/array_action/svc/array_mediator_svc_test.py:721]: its not "custom", its fields are the same, just not one of 1-3.. 
will it be better if I change the `s` to `0`?

Comment (oriyarde) [reply controller/tests/array_action/svc/array_mediator_svc_test.py:721]: > its fields are the same

I thought the point of this special host is to have custom field **values**.
e.g. a "custom" `wwpns_list=['WWNs']` instead of the default `wwpns_list=['wwn3']`

I'm not sure `0` is clearer

Comment (ArbelNathan) [reply controller/tests/array_action/svc/array_mediator_svc_test.py]: ok, done

File: controller/tests/controller_server/utils_test.py [hunk 24]
```diff
@@ -383,51 +385,71 @@ def test_get_volume_id_info_no_id_fail(self):
...
 
-        res = utils.choose_connectivity_type(["iscsi", "fc"])
-        self.assertEqual(res, "fc")
+    def test_choose_connectivity_types(self):
+        nvme = NVME_OVER_FC_CONNECTIVITY_TYPE
+        fc = FC_CONNECTIVITY_TYPE
+        iscsi = ISCSI_CONNECTIVITY_TYPE
+        expected_chosen_by_connectivities_found = {(nvme, fc, iscsi): nvme, (fc, iscsi): fc, (nvme,): nvme, (fc,): fc,
+                                                   (iscsi,): iscsi
+                                                   }
```
Comment (oriyarde) [inline controller/tests/controller_server/utils_test.py:424]: is this the best indentation/formatting?
```suggestion
        expected_chosen_by_connectivities_found = {
            (nvme, fc, iscsi): nvme, (fc, iscsi): fc,
            (nvme,): nvme, (fc,): fc, (iscsi,): iscsi
        }
```

File: controller/tests/array_action/svc/array_mediator_svc_test.py (hunk 20 above)
Comment (oriyarde) [reply controller/tests/array_action/svc/array_mediator_svc_test.py]: could we use "custom" instead of "special" in the parameter name as well?

File: controller/tests/array_action/svc/array_mediator_svc_test.py [hunk 25]
```diff
@@ -621,106 +632,132 @@ def _prepare_mocks_for_get_host_by_identifiers_(self, result_reader_iter, hosts)
...
-        self.assertEqual([config.ISCSI_CONNECTIVITY_TYPE], connectivity_type)
+        self._prepare_mocks_for_get_host_by_identifiers(result_reader_iter)
+        hostname, connectivity_types = self.svc.get_host_by_host_identifiers(
+            Initiators('Test_nqn', ['Test_wwn'], 'iqn.test.2'))
+        self.assertEqual('test_host_2', hostname)
+        self.assertEqual([config.ISCSI_CONNECTIVITY_TYPE], connectivity_types)
+
+    @patch("controller.array_action.svc_cli_result_reader.SVCListResultsReader.__iter__")
+    def test_get_host_by_identifiers_no_other_ports_return_iscsi_host(self, result_reader_iter):
+        host_with_iqn = self._get_host_as_dictionary('host_id_costume', 'test_costume_host',
```
Comment (oriyarde) [inline controller/tests/array_action/svc/array_mediator_svc_test.py:662]:
```suggestion
        host_with_iqn = self._get_host_as_dictionary('costume_host_id', 'test_costume_host',
```

File: controller/tests/controller_server/utils_test.py (hunk 24 above)
Comment (ArbelNathan) [reply controller/tests/controller_server/utils_test.py]: done

File: controller/tests/array_action/svc/array_mediator_svc_test.py (hunk 25 above)
Comment (ArbelNathan) [reply controller/tests/array_action/svc/array_mediator_svc_test.py]: done

File: controller/tests/array_action/svc/array_mediator_svc_test.py (hunk 20 above)
Comment (ArbelNathan) [reply controller/tests/array_action/svc/array_mediator_svc_test.py]: done

**Review by oriyarde — COMMENTED** [review]
LGTM NJ!

**Review by oriyarde — DISMISSED** [review]
